### PR TITLE
Feat/met 5962 debias detected terms storage and report

### PR DESCRIPTION
**Pull request by jeortizquan**

### CODE REVIEW

File: src/main/java/eu/europeana/metis/sandbox/service/debias/ProcessingState.java [hunk 1]
```diff
@@ -42,12 +42,12 @@ public void success(Integer datasetId) {
...
+      DatasetDeBiasEntity datasetDeBiasEntity = datasetDeBiasRepository.findDetectionEntityByDatasetId_DatasetId(datasetId);
+      if (datasetDeBiasEntity == null) {
         fail(datasetId);
         LOGGER.warn("invalid state {} {}", STATE_NAME, datasetId);
       } else {
-        detectRepository.updateState(datasetId, STATE_NAME);
+        datasetDeBiasRepository.updateState(datasetId, STATE_NAME);
         success(datasetId);
         LOGGER.info("success {} {}", STATE_NAME, datasetId);
         // TODO: processing logic here
```
Comment (jochen-vermeulen) [inline src/main/java/eu/europeana/metis/sandbox/service/debias/ProcessingState.java:53]: Is something still missing? This TODO should be evaluated.

File: src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java [hunk 2]
```diff
@@ -421,8 +421,10 @@ public List<LanguageView> getAllLanguages() {
     @ApiResponse(responseCode = "400", description = MESSAGE_FOR_400_CODE)
     @PostMapping(value = "{id}/debias", produces = APPLICATION_JSON_VALUE)
     @ResponseStatus(HttpStatus.OK)
-    public boolean processDebias(@PathVariable("id") Integer datasetId) {
-        return debiasDetectService.process(datasetId);
+    public boolean processDeBias(@PathVariable("id") Integer datasetId) {
+        debiasStateService.cleanDeBiasReport(datasetId);
+        debiasStateService.setState(debiasStateService.getReady());
+        return debiasStateService.process(datasetId);
```
Comment (jochen-vermeulen) [inline src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java:427]: Should we not check whether the dataset is already processed? Or currently processing? In either case, it's probably not a good idea to start again.

File: src/main/java/eu/europeana/metis/sandbox/service/debias/ProcessingState.java [hunk 3]
```diff
@@ -42,12 +42,12 @@ public void success(Integer datasetId) {
...
     try {
-      DetectionEntity detectionEntity = detectRepository.findDetectionEntityByDatasetId_DatasetId(datasetId);
-      if (detectionEntity == null) {
+      DatasetDeBiasEntity datasetDeBiasEntity = datasetDeBiasRepository.findDetectionEntityByDatasetId_DatasetId(datasetId);
+      if (datasetDeBiasEntity == null) {
         fail(datasetId);
         LOGGER.warn("invalid state {} {}", STATE_NAME, datasetId);
       } else {
-        detectRepository.updateState(datasetId, STATE_NAME);
+        datasetDeBiasRepository.updateState(datasetId, STATE_NAME);
```
Comment (jochen-vermeulen) [inline src/main/java/eu/europeana/metis/sandbox/service/debias/ProcessingState.java:50]: Should this not be conditional on detecting that all records have been processed for debias? Or maybe it is, but I'm not seeing it :)

File: src/main/java/eu/europeana/metis/sandbox/dto/debias/DeBiasReportDto.java [hunk 4]
```diff
@@ -2,14 +2,18 @@
...
+import java.util.Collections;
+import java.util.List;
 
 /**
  * The type Detection info dto.
  */
-@ApiModel(DetectionInfoDto.SWAGGER_MODEL_NAME)
-public class DetectionInfoDto {
+@ApiModel(DeBiasReportDto.SWAGGER_MODEL_NAME)
+public class DeBiasReportDto {
```
Comment (jochen-vermeulen) [inline src/main/java/eu/europeana/metis/sandbox/dto/debias/DeBiasReportDto.java:16]: Is the DeBias state also available in another call? It seems inefficient to download the whole report repeatedly just to know whether the report is ready for downloading.

File: src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java (hunk 2 above)
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java:427]: good catch! I'll add the check.

File: src/main/java/eu/europeana/metis/sandbox/dto/debias/DeBiasReportDto.java (hunk 4 above)
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/dto/debias/DeBiasReportDto.java:16]: correct. I'll split this functionality.

File: src/main/java/eu/europeana/metis/sandbox/service/debias/ProcessingState.java (hunk 3 above)
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/service/debias/ProcessingState.java:50]: :thinking: true. I'll refine this process as well.

File: src/main/java/eu/europeana/metis/sandbox/service/debias/ProcessingState.java (hunk 1 above)
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/service/debias/ProcessingState.java:53]: I will do a clean up.

File: src/main/java/eu/europeana/metis/sandbox/entity/debias/RecordDeBiasDetailEntity.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  @JoinColumn(name = "debias_id", referencedColumnName = "id")
+  private RecordDeBiasMainEntity debiasId;
+
+  protected int tag_start;
+
+  protected int tag_end;
+
+  protected int tag_length;
+
+  protected String tag_uri;
```
[the remark this answers is not in the document]
Comment (acerdas84) [reply src/main/java/eu/europeana/metis/sandbox/entity/debias/RecordDeBiasDetailEntity.java:38]: Not sure if this is JPA but translating from camel case to underscores is the default there so variables here could use camel case and still work, but as you mentioned there might be other alternatives like mentioned [here](https://noahhsu.medium.com/spring-jpa-trap-for-camel-case-table-field-name-2fa818e4c3c0)

File: src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java [hunk 6]
```diff
@@ -431,14 +442,30 @@ public boolean processDebias(@PathVariable("id") Integer datasetId) {
      * @param datasetId the dataset id
      * @return the DeBias detection
      */
-    @Operation(description = "Get DeBias detection dataset")
+    @Operation(description = "Get DeBias report detection dataset")
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java:445]: done.

File: src/main/java/eu/europeana/metis/sandbox/dto/debias/DeBiasStatusDto.java [hunk 7]
```diff
@@ -2,19 +2,22 @@
 
 import com.fasterxml.jackson.annotation.JsonFormat;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import eu.europeana.metis.sandbox.service.workflow.DeBiasProcessServiceImpl.DeBiasReportRow;
 import io.swagger.annotations.ApiModel;
 import java.time.ZonedDateTime;
+import java.util.Collections;
+import java.util.List;
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/dto/debias/DeBiasStatusDto.java:9]: done.

File: src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java [hunk 8]
```diff
@@ -421,8 +424,16 @@ public List<LanguageView> getAllLanguages() {
...
+    public boolean processDeBias(@PathVariable("id") Integer datasetId) {
+        ProgressInfoDto progressInfoDto = reportService.getReport(datasetId.toString());
+        if (Optional.ofNullable(debiasStateService.getDeBiasStatus(datasetId))
+                    .map(DeBiasStatusDto::getState)
+                    .orElse("").equals("READY")
+            && progressInfoDto.getStatus().equals(Status.COMPLETED)) {
+            debiasStateService.cleanDeBiasReport(datasetId);
+            return debiasStateService.process(datasetId);
+        } else
+            return false;
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/controller/DatasetController.java:436]: uff. yes. done.

File: src/main/java/eu/europeana/metis/sandbox/entity/debias/RecordDeBiasDetailEntity.java [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+import jakarta.persistence.CascadeType;
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/entity/debias/RecordDeBiasDetailEntity.java:15]: done.

File: src/main/java/eu/europeana/metis/sandbox/entity/debias/RecordDeBiasDetailEntity.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/entity/debias/RecordDeBiasDetailEntity.java:38]: Yes. I is related with JPA, I had issues with psql that `start,end` are reserved words I just added the underscore to the fields, in that iteration to continue working. Now they the fields are in `camelCase`  within the Java class, and internally the framework translates to `tag_xyz`. done.

File: src/main/java/eu/europeana/metis/sandbox/executor/workflow/StepExecutor.java [hunk 10]
```diff
@@ -6,6 +6,7 @@
 import eu.europeana.metis.sandbox.domain.RecordError;
 import eu.europeana.metis.sandbox.domain.RecordInfo;
 import eu.europeana.metis.sandbox.domain.RecordProcessEvent;
+import java.util.ArrayList;
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/executor/workflow/StepExecutor.java:9]: done.

File: src/main/java/eu/europeana/metis/sandbox/repository/RecordLogRepository.java [hunk 11]
```diff
@@ -42,7 +43,7 @@ public interface RecordLogRepository extends JpaRepository<RecordLogEntity, Long
   /**
    * Get record given a record id, dataset id and step.
    * <p>The record id will be searched against both {@link RecordLogEntity#getRecordId().getProviderId()} and {@link
-   * RecordLogEntity#getRecordId().getEuropeanaId()} together with {@link RecordLogEntity#getStep()}.</p>
+   * RecordLogEntity#getRecordId().getEuropeanaId()}* together with {@link RecordLogEntity#getStep()}.</p>
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/RecordLogRepository.java:46]: this strange. probably came from running the intelliJ reformat code function. I just removed the `*`. done.

File: src/main/java/eu/europeana/metis/sandbox/repository/RecordLogRepository.java [hunk 12]
```diff
@@ -57,7 +58,7 @@ public interface RecordLogRepository extends JpaRepository<RecordLogEntity, Long
   /**
    * Get set of records given a record id, dataset id and steps.
    * <p>The record id will be searched against both {@link RecordLogEntity#getRecordId().getProviderId()} and {@link
-   * RecordLogEntity#getRecordId().getEuropeanaId()} together with {@link RecordLogEntity#getStep()}.</p>
+   * RecordLogEntity#getRecordId().getEuropeanaId()}* together with {@link RecordLogEntity#getStep()}.</p>
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/RecordLogRepository.java:61]: same here. done.

File: src/main/java/eu/europeana/metis/sandbox/repository/RecordLogRepository.java [hunk 13]
```diff
@@ -68,11 +69,36 @@ public interface RecordLogRepository extends JpaRepository<RecordLogEntity, Long
...
   void deleteByRecordIdDatasetId(String datasetId);
+
+  /**
+   * Delete by record id dataset id and step.
+   *
+   * @param datasetId the dataset id
+   * @param step the step
+   */
+  @Modifying
+  void deleteByRecordIdDatasetIdAndStep(String datasetId, Step step);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/RecordLogRepository.java:103]: This weird naming convention that has to do with the JPA framework that is how it detects record has a datasetId and step :shrug: if I change the name the queries stop working.

File: src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasServiceable.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
+package eu.europeana.metis.sandbox.service.debias;
+
+import eu.europeana.metis.sandbox.dto.debias.DeBiasReportDto;
+import eu.europeana.metis.sandbox.dto.debias.DeBiasStatusDto;
+
+
+/**
+ * The interface DeBias stateful.
+ */
+public interface DeBiasServiceable {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasServiceable.java:10]: It is a matter of taste I prefer to use `adjectives` for the interfaces and `nouns` for the classes. Unfortunately all the code has this `Service` `ServiceImpl` naming convention.

[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasServiceable.java:10]: Changed to the old convention naming. done. We can create a ticket to use a different naming convention later.

File: src/main/java/eu/europeana/metis/sandbox/repository/debias/RecordDeBiasDetailRepository.java [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+   */
+  List<RecordDeBiasDetailEntity> findByDebiasIdId(Long debiasId);
+
+  /**
+   * Delete by debias id record id dataset id.
+   *
+   * @param datasetId the dataset id
+   */
+  @Modifying
+  void deleteByDebiasIdRecordIdDatasetId(String datasetId);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/debias/RecordDeBiasDetailRepository.java:28]: yes. is again the JPA framework naming convention to do the queries.

File: src/main/java/eu/europeana/metis/sandbox/repository/debias/RecordDeBiasMainRepository.java [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+public interface RecordDeBiasMainRepository extends JpaRepository<RecordDeBiasMainEntity, Long> {
+
+
+  /**
+   * Find by record id dataset id list.
+   *
+   * @param datasetId the dataset id
+   * @return the list
+   */
+  List<RecordDeBiasMainEntity> findByRecordIdDatasetId(String datasetId);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/debias/RecordDeBiasMainRepository.java:21]: yes. same here is JPA framework naming convention to do the queries.

File: src/main/java/eu/europeana/metis/sandbox/repository/debias/RecordDeBiasMainRepository.java [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+
+  /**
+   * Delete by record id dataset id.
+   *
+   * @param datasetId the dataset id
+   */
+  @Modifying
+  void deleteByRecordIdDatasetId(String datasetId);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/debias/RecordDeBiasMainRepository.java:29]: also here JPA framework naming convention.

File: src/main/java/eu/europeana/metis/sandbox/repository/debias/DatasetDeBiasRepository.java [hunk 18]
```diff
@@ -1,22 +1,23 @@
...
+public interface DatasetDeBiasRepository extends JpaRepository<DatasetDeBiasEntity, Long> {
 
   /**
    * Find detection entity by dataset id detection entity.
    *
    * @param datasetId the dataset id
    * @return the detection entity
    */
-  DetectionEntity findDetectionEntityByDatasetId_DatasetId(Integer datasetId);
+  DatasetDeBiasEntity findDetectionEntityByDatasetId_DatasetId(Integer datasetId);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/debias/DatasetDeBiasRepository.java:20]: is JPA framework naming convention, also is possible to remove the underscore, now it looks like this `findDetectionEntityByDatasetIdDatasetId` :see_no_evil: done.

File: src/main/java/eu/europeana/metis/sandbox/service/workflow/DeBiasProcessService.java [hunk 19]
```diff
@@ -0,0 +1,18 @@
+package eu.europeana.metis.sandbox.service.workflow;
+
+import eu.europeana.metis.sandbox.domain.Record;
+import eu.europeana.metis.sandbox.domain.RecordInfo;
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/service/workflow/DeBiasProcessService.java:4]: done.

File: src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasStateService.java [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+  public DeBiasStatusDto getDeBiasStatus(Integer datasetId) {
+    DatasetDeBiasEntity datasetDeBiasEntity = datasetDeBiasRepository.findDetectionEntityByDatasetId_DatasetId(datasetId);
+    if (datasetDeBiasEntity == null) {
+      return new DeBiasStatusDto(datasetId, STATE_NAME, ZonedDateTime.now());
+    } else {
+      return new DeBiasStatusDto(datasetId, datasetDeBiasEntity.getState(), datasetDeBiasEntity.getCreatedDate());
+    }
+  }
+
+  private @NotNull DatasetEntity getDatasetAndProcessDetectionEntity(Integer datasetId) {
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasStateService.java:130]: yes. the naming was a leftover from a previous implementation. I renamed it to what it does now. done.

File: src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasStateService.java [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+    DatasetEntity dataset = datasetRepository.findById(datasetId).orElseThrow();
+    DatasetDeBiasEntity datasetDeBiasEntity = datasetDeBiasRepository.findDetectionEntityByDatasetId_DatasetId(datasetId);
+    if (datasetDeBiasEntity == null) {
+      datasetDeBiasEntity = new DatasetDeBiasEntity(dataset, STATE_NAME);
+      datasetDeBiasRepository.save(datasetDeBiasEntity);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasStateService.java:135]: done.

File: src/main/java/eu/europeana/metis/sandbox/repository/debias/RecordDeBiasDetailRepository.java [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+public interface RecordDeBiasDetailRepository extends JpaRepository<RecordDeBiasDetailEntity, Long> {
+
+
+  /**
+   * Find by debias id list.
+   *
+   * @param debiasId the debias id
+   * @return the list
+   */
+  List<RecordDeBiasDetailEntity> findByDebiasIdId(Long debiasId);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/debias/RecordDeBiasDetailRepository.java:20]: yes. is JPA naming convention for queries.

File: src/main/java/eu/europeana/metis/sandbox/repository/RecordLogRepository.java [hunk 23]
```diff
@@ -68,11 +69,36 @@ public interface RecordLogRepository extends JpaRepository<RecordLogEntity, Long
...
+   */
+  @Query("SELECT rle FROM RecordLogEntity rle WHERE rle.recordId.datasetId = ?1 AND rle.step = ?2")
+  Set<RecordLogEntity> findRecordLogByDatasetIdAndStep(String datasetId, Step step);
+
+  @Query("SELECT rle FROM RecordLogEntity rle WHERE rle.recordId.datasetId = ?1 AND rle.step = ?2 AND rle.status = ?3")
+  Set<RecordLogEntity> findRecordLogByDatasetIdAndStepAndStatus(String datasetId, Step step, Status status);
+
+  @Modifying
+  @Query("UPDATE RecordLogEntity rle SET rle.status = ?3 WHERE rle.recordId.id = ?1 AND rle.step = ?2")
+  void updateByRecordIdAndStepAndStatus(Long recordId, Step step, Status status);
```
[the remark this answers is not in the document]
Comment (jeortizquan) [reply src/main/java/eu/europeana/metis/sandbox/repository/RecordLogRepository.java:87]: I am using JPA convention style for the method. that is the reason for this odd names.

File: src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasServiceable.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (jochen-vermeulen) [reply src/main/java/eu/europeana/metis/sandbox/service/debias/DeBiasServiceable.java]: OK. not saying one is better than the other.

**Pull request merged**